ReactiveCocoa/ObjC+MessageForwarding.swift

```diff
@@ -0,0 +1,249 @@
+		let selector = invocation.selector!
+		let alias = selectorCache.alias(for: selector)
+		let interopAlias = selectorCache.interopAlias(for: selector)
+
+		defer { interceptingObject(objectRef, didInvokeSelectorOfAlias: alias, with: invocation) }
+
+		if let method = class_getInstanceMethod(perceivedClass, selector) {
+			let typeEncoding = method_getTypeEncoding(method)
+
+			if class_respondsToSelector(realClass, interopAlias) {
+				// RAC has preserved an immediate implementation found in the runtime
+				// subclass that was supplied by an external party.
+				//
+				// As the KVO setter relies on the selector to work, it has to be invoked
+				// by swapping in the preserved implementation and restore to the message
+				// forwarder afterwards.
+				//
+				// However, the IMP cache would be thrashed due to the swapping.
+
+				let topLevelClass: AnyClass = object_getClass(objectRef.takeUnretainedValue())
+
+				// The locking below prevents RAC swizzling attempts from intervening the
+				// invocation.
+				//
+				// Given the implementation of `swizzleClass`, `topLevelClass` can only be:
+				// (1) the same as `realClass`; or (2) a subclass of `realClass`. In other
+				// words, this would deadlock only if the locking order is not followed in
+				// other nested locking scenarios of these metaclasses at compile time.
+
+				synchronized(topLevelClass) {
+					func swizzle() {
+						let interopImpl = class_getMethodImplementation(topLevelClass, interopAlias)
+
+						let previousImpl = class_replaceMethod(topLevelClass, selector, interopImpl, typeEncoding)
+						invocation.invoke()
+
+						_ = class_replaceMethod(topLevelClass, selector, previousImpl, typeEncoding)
+					}
+
+					if topLevelClass != realClass {
+						synchronized(realClass) {
+							// In addition to swapping in the implementation, the message
+							// forwarding needs to be temporarily disabled to prevent circular
+							// invocation.
+							_ = class_replaceMethod(realClass, selector, nil, typeEncoding)
+							swizzle()
+							_ = class_replaceMethod(realClass, selector, _rac_objc_msgForward, typeEncoding)
+						}
+					} else {
+						swizzle()
+					}
+				}
+
+				return
+			}
+
+			if let impl = method_getImplementation(method), impl != _rac_objc_msgForward {
+				// The perceived class, or its ancestors, responds to the selector.
+				//
+				// The implementation is invoked through the selector alias, which
+				// reflects the latest implementation of the selector in the perceived
+				// class.
+
+				if class_getMethodImplementation(realClass, alias) != impl {
+					// Update the alias if and only if the implementation has changed, so as
+					// to avoid thrashing the IMP cache.
+					_ = class_replaceMethod(realClass, alias, impl, typeEncoding)
+				}
+
+				invocation.setSelector(alias)
+				invocation.invoke()
+
+				return
+			}
+		}
+
+		guard !isDelegateProxy else {
```

Comment:
NOTE: Perhaps the DelegateProxy would be better to have its own simplified version of `forwardInvocation`. Much of the code is useless for the proxy, or can be assumed non-interoperable as we own the proxy type. 🤔